Migrate/handle set item name change

https://github.com/aikaterna/gobcog/commit/fd360fc5400ee9813dde3ad2b36557f1e26f415a introduced an issue because I didn't migrate things
```
[2020-06-24 22:56:23] [ERROR] red: Exception in command 'equip'
Traceback (most recent call last):
  File "/.../cat38/lib/python3.8/site-packages/discord/ext/commands/core.py", line 83, in wrapped
    ret = await coro(*args, **kwargs)
  File "/.../Red-DiscordBot/cogs/CogManager/cogs/adventure/adventure.py", line 2425, in equip
    await ctx.invoke(self.backpack_equip, equip_item=item)
  File "/.../cat38/lib/python3.8/site-packages/discord/ext/commands/context.py", line 132, in invoke
    ret = await command.callback(*arguments, **kwargs)
  File "/.../Red-DiscordBot/cogs/CogManager/cogs/adventure/adventure.py", line 838, in backpack_equip
    equip = c.backpack[equip_item.name]
KeyError: 'King Solomons Shoes'
```
```
[2020-06-24 16:43:23] [ERROR] red: Exception in command 'equip'
Traceback (most recent call last):
  File "/.../cat38/lib/python3.8/site-packages/discord/ext/commands/core.py", line 83, in wrapped
    ret = await coro(*args, **kwargs)
  File "/.../Red-DiscordBot/cogs/CogManager/cogs/adventure/adventure.py", line 2393, in equip
    await ctx.invoke(self.backpack_equip, equip_item=item)
  File "/.../cat38/lib/python3.8/site-packages/discord/ext/commands/context.py", line 132, in invoke
    ret = await command.callback(*arguments, **kwargs)
  File "/.../Red-DiscordBot/cogs/CogManager/cogs/adventure/adventure.py", line 818, in backpack_equip
    equip = c.backpack[equip_item.name]
KeyError: 'King Solomons Upper Robe'
```

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -3813,7 +3813,6 @@ class Adventure(BaseCog):
                                 else _("1 second")
                             ),
                         )
-                    ainz = False
                     theme = await self.config.theme()
                     extra_pets = await self.config.themes.all()
                     extra_pets = extra_pets.get(theme, {}).get("pets", {})
@@ -3825,18 +3824,27 @@ class Adventure(BaseCog):
                     pet_reqs = pet_list[pet].get("bonuses", {}).get("req", {})
                     pet_msg4 = ""
                     can_catch = True
-                    if pet_reqs.get("set", False):
+                    force_catch = False
+                    if "Ainz Ooal Gown" in c.sets:
+                        can_catch = True
+                        pet = random.choice(
+                            [
+                                "Albedo",
+                                "Rubedo",
+                                "Guardians of Nazarick",
+                                *random.choices(pet_choices, k=10),
+                            ]
+                        )
+                        if pet in ["Albedo", "Rubedo", "Guardians of Nazarick"]:
+                            force_catch = True
+                    elif pet_reqs.get("bonuses", {}).get("req"):
                         if pet_reqs.get("set", None) in c.sets:
                             can_catch = True
-                            if pet_reqs.get("set", None) in ["Ainz Ooal Gown"]:
-                                ainz = True
-                                pet = random.choice(["Albedo", "Rubedo", "Guardians of Nazarick"])
                         else:
                             can_catch = False
                             pet_msg4 = _(
                                 "\nPerhaps you're missing some requirements to tame {pet}."
                             ).format(pet=pet)
-
                     pet_msg = box(
                         _("{c} is trying to tame a pet.").format(
                             c=self.escape(ctx.author.display_name)
@@ -3863,22 +3871,40 @@ class Adventure(BaseCog):
                         bonus = _("But they stepped on a twig and scared it away.")
                     elif roll in [50, 25]:
                         bonus = _("They happen to have its favorite food.")
-                    if ainz is True or (
-                        dipl_value > self.PETS[pet]["cha"] and roll > 1 and can_catch
+                    if force_catch is True or (
+                        dipl_value > pet_list[pet]["cha"] and roll > 1 and can_catch
                     ):
-                        if ainz:
+                        if force_catch:
                             roll = 0
                         else:
                             roll = random.randint(0, 2 if roll in [50, 25] else 5)
                         if roll == 0:
-                            pet_msg3 = box(
-                                _("{bonus}\nThey successfully tamed the {pet}.").format(
-                                    bonus=bonus, pet=pet
-                                ),
-                                lang="css",
-                            )
+                            if force_catch and "Ainz Ooal Gown" in c.sets:
+                                msg = random.choice(
+                                    [
+                                        _("{author} commands {pet} into submission.").format(
+                                            pet=pet, author=self.escape(ctx.author.display_name)
+                                        ),
+                                        _("{pet} swears allegiance to the Supreme One.").format(
+                                            pet=pet, author=self.escape(ctx.author.display_name)
+                                        ),
+                                        _(
+                                            "{pet} takes an Oath of Allegiance to the Supreme One."
+                                        ).format(
+                                            pet=pet, author=self.escape(ctx.author.display_name)
+                                        ),
+                                    ]
+                                )
+                                pet_msg3 = box(msg, lang="css",)
+                            else:
+                                pet_msg3 = box(
+                                    _("{bonus}\nThey successfully tamed the {pet}.").format(
+                                        bonus=bonus, pet=pet
+                                    ),
+                                    lang="css",
+                                )
                             await user_msg.edit(content=f"{pet_msg}\n{pet_msg2}\n{pet_msg3}")
-                            c.heroclass["pet"] = self.PETS[pet]
+                            c.heroclass["pet"] = pet_list[pet]
                             c.heroclass["catch_cooldown"] = time.time()
                             await self.config.user(ctx.author).set(await c.to_json(self.config))
                         elif roll == 1:

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -3455,11 +3455,11 @@ class Adventure(BaseCog):
                     rjust = max([len(str(i)) for i in items.values()])
                     async for item in AsyncIter(items.values()):
                         settext = ""
-                        att_space = " " if len(str(item.att)) == 1 else ""
-                        cha_space = " " if len(str(item.cha)) == 1 else ""
-                        int_space = " " if len(str(item.int)) == 1 else ""
-                        dex_space = " " if len(str(item.dex)) == 1 else ""
-                        luck_space = " " if len(str(item.luck)) == 1 else ""
+                        att_space = " " if len(str(item.att)) >= 1 else ""
+                        cha_space = " " if len(str(item.cha)) >= 1 else ""
+                        int_space = " " if len(str(item.int)) >= 1 else ""
+                        dex_space = " " if len(str(item.dex)) >= 1 else ""
+                        luck_space = " " if len(str(item.luck)) >= 1 else ""
                         owned = f" | {item.owned}"
                         if item.set:
                             settext += f" | Set `{item.set}` ({item.parts}pcs)"

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -907,7 +907,8 @@ class Adventure(BaseCog):
                         level=equiplevel
                     ),
                 )
-            equip = c.backpack[equip_item.name]
+
+            equip = c.backpack.get(equip_item.name)
             if equip:
                 slot = equip.slot[0]
                 if len(equip.slot) > 1:

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -68,7 +68,7 @@ log = logging.getLogger("red.cogs.adventure")
 
 REBIRTH_LVL = 20
 REBIRTH_STEP = 10
-_SCHEMA_VERSION = 3
+_SCHEMA_VERSION = 4
 _config: Config = None
 
 
@@ -577,6 +577,40 @@ class Adventure(BaseCog):
                     except Exception:
                         adventurers_data[user]["loadouts"] = {}
             await self.config.schema_version.set(3)
+
+        if from_version < 4 <= to_version:
+            group = self.config._get_base_group(self.config.USER)
+            accounts = await group.all()
+            tmp = accounts.copy()
+            async with group.all() as adventurers_data:
+                async for user in AsyncIter(tmp, steps=10):
+                    equipped = tmp[user]["items"]
+                    for slot, item in equipped.items():
+                        for item_name, item_data in item.items():
+                            if "King Solomos" in item_name:
+                                del adventurers_data[user]["items"][slot][item_name]
+                                item_name = item_name.replace("Solomos", "Solomons")
+                                adventurers_data[user]["items"][slot][item_name] = item_data
+                    loadout = tmp[user]["loadouts"]
+                    for loadout_name, loadout_data in loadout.items():
+                        for slot, item in equipped.items():
+                            for item_name, item_data in item.items():
+                                if "King Solomos" in item_name:
+                                    del adventurers_data[user]["loadouts"][loadout_name][slot][
+                                        item_name
+                                    ]
+                                    item_name = item_name.replace("Solomos", "Solomons")
+                                    adventurers_data[user]["loadouts"][loadout_name][slot][
+                                        item_name
+                                    ] = item_data
+
+                    backpack = tmp[user]["backpack"]
+                    async for item_name, item_data in AsyncIter(backpack.items(), steps=25):
+                        if "King Solomos" in item_name:
+                            del adventurers_data[user]["backpack"][item_name]
+                            item_name = item_name.replace("Solomos", "Solomons")
+                            adventurers_data[user]["backpack"][item_name] = item_data
+            await self.config.schema_version.set(4)
 
     def _convert_item_migration(self, item_name, item_dict):
         new_name = item_name

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -770,6 +770,7 @@ class Adventure(BaseCog):
     async def _backpack(
         self,
         ctx: Context,
+        show_diff: Optional[bool] = False,
         rarity: Optional[RarityConverter] = None,
         *,
         slot: Optional[SlotConverter] = None,
@@ -817,7 +818,7 @@ class Adventure(BaseCog):
 
             backpack_contents = _("[{author}'s backpack] \n\n{backpack}\n").format(
                 author=self.escape(ctx.author.display_name),
-                backpack=await c.get_backpack(rarity=rarity, slot=slot),
+                backpack=await c.get_backpack(rarity=rarity, slot=slot, show_delta=show_diff),
             )
             msgs = []
             async for page in AsyncIter(

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -3813,6 +3813,7 @@ class Adventure(BaseCog):
                                 else _("1 second")
                             ),
                         )
+                    ainz = False
                     theme = await self.config.theme()
                     extra_pets = await self.config.themes.all()
                     extra_pets = extra_pets.get(theme, {}).get("pets", {})
@@ -3827,6 +3828,9 @@ class Adventure(BaseCog):
                     if pet_reqs.get("set", False):
                         if pet_reqs.get("set", None) in c.sets:
                             can_catch = True
+                            if pet_reqs.get("set", None) in ["Ainz Ooal Gown"]:
+                                ainz = True
+                                pet = random.choice(["Albedo", "Rubedo", "Guardians of Nazarick"])
                         else:
                             can_catch = False
                             pet_msg4 = _(
@@ -3859,8 +3863,13 @@ class Adventure(BaseCog):
                         bonus = _("But they stepped on a twig and scared it away.")
                     elif roll in [50, 25]:
                         bonus = _("They happen to have its favorite food.")
-                    if dipl_value > self.PETS[pet]["cha"] and roll > 1 and can_catch:
-                        roll = random.randint(0, 2 if roll in [50, 25] else 5)
+                    if ainz is True or (
+                        dipl_value > self.PETS[pet]["cha"] and roll > 1 and can_catch
+                    ):
+                        if ainz:
+                            roll = 0
+                        else:
+                            roll = random.randint(0, 2 if roll in [50, 25] else 5)
                         if roll == 0:
                             pet_msg3 = box(
                                 _("{bonus}\nThey successfully tamed the {pet}.").format(

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -6397,14 +6397,19 @@ class Adventure(BaseCog):
             adv_end = await self._get_epoch(secondint)
             timer, done, sremain = await self._remaining(adv_end)
             message_adv = await ctx.send(f"⏳ [{title}] {timer}s")
+            deleted = False
             while not done:
                 timer, done, sremain = await self._remaining(adv_end)
                 self._adventure_countdown[ctx.guild.id] = (timer, done, sremain)
                 if done:
-                    await message_adv.delete()
+                    if not deleted:
+                        await message_adv.delete()
                     break
-                elif int(sremain) % 5 == 0:
-                    await message_adv.edit(content=f"⏳ [{title}] {timer}s")
+                elif not deleted and int(sremain) % 5 == 0:
+                    try:
+                        await message_adv.edit(content=f"⏳ [{title}] {timer}s")
+                    except discord.NotFound:
+                        deleted = True
                 await asyncio.sleep(1)
             log.debug("Timer countdown done.")
 
@@ -6419,14 +6424,19 @@ class Adventure(BaseCog):
             cart_end = await self._get_epoch(secondint)
             timer, done, sremain = await self._remaining(cart_end)
             message_cart = await room.send(f"⏳ [{title}] {timer}s")
+            deleted = False
             while not done:
                 timer, done, sremain = await self._remaining(cart_end)
                 self._trader_countdown[ctx.guild.id] = (timer, done, sremain)
                 if done:
-                    await message_cart.delete()
+                    if not deleted:
+                        await message_cart.delete()
                     break
-                if int(sremain) % 5 == 0:
-                    await message_cart.edit(content=f"⏳ [{title}] {timer}s")
+                if not deleted and int(sremain) % 5 == 0:
+                    try:
+                        await message_cart.edit(content=f"⏳ [{title}] {timer}s")
+                    except discord.NotFound:
+                        deleted = True
                 await asyncio.sleep(1)
 
         return ctx.bot.loop.create_task(cart_countdown())

--- a/adventure/adventure.py
+++ b/adventure/adventure.py
@@ -216,7 +216,7 @@ class AdventureResults:
 class Adventure(BaseCog):
     """Adventure, derived from the Goblins Adventure cog by locastan."""
 
-    __version__ = "3.2.27"
+    __version__ = "3.2.28"
 
     def __init__(self, bot: Red):
         self.bot = bot

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -572,7 +572,7 @@ class Character(Item):
                 parts, count, bonus = set_names[item.set]
                 set_names[item.set] = (parts, count + 1, bonus)
         valid_sets = [(s, v[1]) for s, v in set_names.items() if v[1] >= v[0]]
-        self.sets = [s for s, _ in set_names.items() if s]
+        self.sets = [s for s, _ in valid_sets if s]
         for (_set, parts) in valid_sets:
             set_bonuses = SET_BONUSES.get(_set, [])
             for bonus in set_bonuses:

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -852,13 +852,15 @@ class Character(Item):
                     int = self.get_equipped_delta(current_equipped, item[1], "int")
                     dex = self.get_equipped_delta(current_equipped, item[1], "dex")
                     luck = self.get_equipped_delta(current_equipped, item[1], "luck")
+                    rjuststat = 5
                 else:
                     att = item[1].att if len(slot_name_org) < 2 else item[1].att * 2
                     cha = item[1].cha if len(slot_name_org) < 2 else item[1].cha * 2
                     int = item[1].int if len(slot_name_org) < 2 else item[1].int * 2
                     dex = item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2
                     luck = item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2
-                rjuststat = 5
+                    rjuststat = 3
+
                 stats = (
                     f"({att_space}{att:<{rjuststat}} |"
                     f"{cha_space}{cha:<{rjuststat}} |"

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -605,9 +605,18 @@ class Character(Item):
                 if not self.heroclass["pet"]:
                     class_desc += _("\n\n- Current pet: [None]")
                 elif self.heroclass["pet"]:
-                    class_desc += _("\n\n- Current pet: [{}]").format(
-                        self.heroclass["pet"]["name"]
-                    )
+                    if "Ainz Ooal Gown" in self.sets and self.heroclass["pet"]["name"] in [
+                        "Albedo",
+                        "Rubedo",
+                        "Guardians of Nazarick",
+                    ]:
+                        class_desc += _("\n\n- Current servant: [{}]").format(
+                            self.heroclass["pet"]["name"]
+                        )
+                    else:
+                        class_desc += _("\n\n- Current pet: [{}]").format(
+                            self.heroclass["pet"]["name"]
+                        )
         else:
             class_desc = _("Hero.")
 
@@ -1264,10 +1273,28 @@ class ItemConverter(Converter):
         except Exception as exc:
             log.exception("Error with the new character sheet", exc_info=exc)
             raise BadArgument
+        equipped_items = set()
+        for slots in ORDER:
+            if slots == "two handed":
+                continue
+            item = getattr(c, slots, None)
+            if item:
+                equipped_items.add(str(item))
         no_markdown = Item.remove_markdowns(argument)
-        lookup = list(i for x, i in c.backpack.items() if no_markdown.lower() in x.lower())
-        lookup_m = list(i for x, i in c.backpack.items() if argument.lower() == str(i).lower())
-        lookup_e = list(i for x, i in c.backpack.items() if argument == str(i))
+        lookup = list(
+            i
+            for x, i in c.backpack.items()
+            if no_markdown.lower() in x.lower() and str(i) not in equipped_items
+        )
+        lookup_m = list(
+            i
+            for x, i in c.backpack.items()
+            if argument.lower() == str(i).lower() and str(i) not in equipped_items
+        )
+        lookup_e = list(
+            i for x, i in c.backpack.items() if argument == str(i) and str(i) not in equipped_items
+        )
+
         _temp_items = set()
         for i in lookup:
             _temp_items.add(str(i))

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -772,7 +772,7 @@ class Character(Item):
         else:
             return 6  # common / normal
 
-    async def get_sorted_backpack(self, backpack: dict):
+    async def get_sorted_backpack(self, backpack: dict, slot=None, rarity=None):
         tmp = {}
 
         def _sort(item):
@@ -783,6 +783,10 @@ class Character(Item):
             slot_name = slots[0]
             if len(slots) > 1:
                 slot_name = "two handed"
+            if slot is not None and slot_name != slot:
+                continue
+            if rarity is not None and rarity != backpack[item].rarity:
+                continue
 
             if slot_name not in tmp:
                 tmp[slot_name] = []
@@ -790,7 +794,8 @@ class Character(Item):
 
         final = []
         async for (idx, slot_name) in AsyncIter(tmp.keys()).enumerate():
-            final.append(sorted(tmp[slot_name], key=_sort))
+            if tmp[slot_name]:
+                final.append(sorted(tmp[slot_name], key=_sort))
 
         final.sort(
             key=lambda i: ORDER.index(i[0][1].slot[0])
@@ -804,7 +809,7 @@ class Character(Item):
     ):
         if consumed is None:
             consumed = []
-        bkpk = await self.get_sorted_backpack(self.backpack)
+        bkpk = await self.get_sorted_backpack(self.backpack, slot=slot, rarity=rarity)
         form_string = _(
             "Items in Backpack: \n( ATT | CHA | INT | DEX | LUCK ) | LEVEL REQ | [DEGRADE#] | OWNED | SET (SET PIECES)"
         )
@@ -847,23 +852,12 @@ class Character(Item):
                     int = self.get_equipped_delta(current_equipped, item[1], "int")
                     dex = self.get_equipped_delta(current_equipped, item[1], "dex")
                     luck = self.get_equipped_delta(current_equipped, item[1], "luck")
-                    rjuststat = max([len(att), len(cha), len(int), len(dex), len(luck)])
-
                 else:
                     att = item[1].att if len(slot_name_org) < 2 else item[1].att * 2
                     cha = item[1].cha if len(slot_name_org) < 2 else item[1].cha * 2
                     int = item[1].int if len(slot_name_org) < 2 else item[1].int * 2
                     dex = item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2
                     luck = item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2
-                    rjuststat = max(
-                        [
-                            len(str(att)),
-                            len(str(cha)),
-                            len(str(int)),
-                            len(str(dex)),
-                            len(str(luck)),
-                        ]
-                    )
                 rjuststat = 5
                 stats = (
                     f"({att_space}{att:<{rjuststat}} |"

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -718,7 +718,7 @@ class Character(Item):
             luck_space = " " if len(str(luck)) == 1 else ""
 
             owned = ""
-            if item.rarity in ["legendary", "event"] and item.degrade > 0:
+            if item.rarity in ["legendary", "event"] and item.degrade >= 0:
                 owned += f" | [{item.degrade}#]"
             if item.set:
                 settext += f" | Set `{item.set}` ({item.parts}pcs)"
@@ -800,7 +800,7 @@ class Character(Item):
         )
         return final
 
-    async def get_backpack(self, forging: bool = False, consumed=None, rarity=None, slot=None):
+    async def get_backpack(self, forging: bool = False, consumed=None, rarity=None, slot=None, show_delta=False):
         if consumed is None:
             consumed = []
         bkpk = await self.get_sorted_backpack(self.backpack)
@@ -828,7 +828,7 @@ class Character(Item):
                 dex_space = " " if len(str(item[1].dex)) == 1 else ""
                 luck_space = " " if len(str(item[1].luck)) == 1 else ""
                 owned = ""
-                if item[1].rarity in ["legendary", "event"] and item[1].degrade > 0:
+                if item[1].rarity in ["legendary", "event"] and item[1].degrade >= 0:
                     owned += f" | [{item[1].degrade}#]"
                 owned += f" | {item[1].owned}"
                 if item[1].set:
@@ -839,13 +839,18 @@ class Character(Item):
                 else:
                     level = f"{e_level}"
 
-                form_string += (
-                    f"\n{str(item[1]):<{rjust}} - "
-                    f"({att_space}{item[1].att if len(slot_name_org)  < 2 else item[1].att * 2} |"
+                if show_delta:
+                    pass
+                else:
+                    stats = (f"({att_space}{item[1].att if len(slot_name_org)  < 2 else item[1].att * 2} |"
                     f"{cha_space}{item[1].cha if len(slot_name_org)  < 2 else item[1].cha * 2} |"
                     f"{int_space}{item[1].int if len(slot_name_org) < 2 else item[1].int * 2} |"
                     f"{dex_space}{item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2} |"
-                    f"{luck_space}{item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2} )"
+                    f"{luck_space}{item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2} )")
+
+                form_string += (
+                    f"\n{str(item[1]):<{rjust}} - "
+                    f"{stats}"
                     f" | Lvl {level:<5}"
                     f"{owned}{settext}"
                 )
@@ -1139,7 +1144,7 @@ class Character(Item):
                 elif self.rebirths < 50 and i.get("rarity", False) in ["legendary", "event"]:
                     if "degrade" in i:
                         i["degrade"] -= 1
-                        if i.get("degrade", 0) >= 1:
+                        if i.get("degrade", 0) >= 0:
                             backpack[n] = i
 
         tresure = [0, 0, 0, 0, 0]

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -1,7 +1,6 @@
 # -*- coding: utf-8 -*-
 import asyncio
 import logging
-import operator
 import re
 from copy import copy
 from datetime import date, timedelta, datetime
@@ -649,11 +648,11 @@ class Character(Item):
             else humanize_number(max_level_xp),
             skill_points=0 if self.skill["pool"] < 0 else self.skill["pool"],
             set_bonus=(
-                f"( {self.gear_set_bonus.get('att')} | "
-                f"{self.gear_set_bonus.get('cha')} | "
-                f"{self.gear_set_bonus.get('int')} | "
-                f"{self.gear_set_bonus.get('dex')} | "
-                f"{self.gear_set_bonus.get('luck')} ) "
+                f"( {self.gear_set_bonus.get('att'):<2} | "
+                f"{self.gear_set_bonus.get('cha'):<2} | "
+                f"{self.gear_set_bonus.get('int'):<2} | "
+                f"{self.gear_set_bonus.get('dex'):<2} | "
+                f"{self.gear_set_bonus.get('luck'):<2} ) "
                 f"Stats: {round(statmult * 100)}% | "
                 f"EXP: {round(xpmult * 100)}% | "
                 f"Credits: {round(cpmult * 100)}%"
@@ -711,11 +710,11 @@ class Character(Item):
                     * self.gear_set_bonus.get("statmult", 1)
                 )
             )
-            att_space = " " if len(str(att)) == 1 else ""
-            cha_space = " " if len(str(cha)) == 1 else ""
-            int_space = " " if len(str(inter)) == 1 else ""
-            dex_space = " " if len(str(dex)) == 1 else ""
-            luck_space = " " if len(str(luck)) == 1 else ""
+            att_space = " " if len(str(att)) >= 1 else ""
+            cha_space = " " if len(str(cha)) >= 1 else ""
+            int_space = " " if len(str(inter)) >= 1 else ""
+            dex_space = " " if len(str(dex)) >= 1 else ""
+            luck_space = " " if len(str(luck)) >= 1 else ""
 
             owned = ""
             if item.rarity in ["legendary", "event"] and item.degrade >= 0:
@@ -724,11 +723,11 @@ class Character(Item):
                 settext += f" | Set `{item.set}` ({item.parts}pcs)"
             form_string += (
                 f"\n{str(item):<{rjust}} - "
-                f"({att_space}{att} |"
-                f"{cha_space}{cha} |"
-                f"{int_space}{inter} |"
-                f"{dex_space}{dex} |"
-                f"{luck_space}{luck} )"
+                f"({att_space}{att:<3} |"
+                f"{cha_space}{cha:<3} |"
+                f"{int_space}{inter:<3} |"
+                f"{dex_space}{dex:<3} |"
+                f"{luck_space}{luck:<3} )"
                 f" | Lvl { equip_level(self, item):<5}"
                 f"{owned}{settext}"
             )
@@ -825,11 +824,11 @@ class Character(Item):
                     continue
 
                 settext = ""
-                att_space = " " if len(str(item[1].att)) == 1 else ""
-                cha_space = " " if len(str(item[1].cha)) == 1 else ""
-                int_space = " " if len(str(item[1].int)) == 1 else ""
-                dex_space = " " if len(str(item[1].dex)) == 1 else ""
-                luck_space = " " if len(str(item[1].luck)) == 1 else ""
+                att_space = " " if len(str(item[1].att)) >= 1 else ""
+                cha_space = " " if len(str(item[1].cha)) >= 1 else ""
+                int_space = " " if len(str(item[1].int)) >= 1 else ""
+                dex_space = " " if len(str(item[1].dex)) >= 1 else ""
+                luck_space = " " if len(str(item[1].luck)) >= 1 else ""
                 owned = ""
                 if item[1].rarity in ["legendary", "event"] and item[1].degrade >= 0:
                     owned += f" | [{item[1].degrade}#]"
@@ -843,22 +842,36 @@ class Character(Item):
                     level = f"{e_level}"
 
                 if show_delta:
+                    att = self.get_equipped_delta(current_equipped, item[1], "att")
+                    cha = self.get_equipped_delta(current_equipped, item[1], "cha")
+                    int = self.get_equipped_delta(current_equipped, item[1], "int")
+                    dex = self.get_equipped_delta(current_equipped, item[1], "dex")
+                    luck = self.get_equipped_delta(current_equipped, item[1], "luck")
+                    rjuststat = max([len(att), len(cha), len(int), len(dex), len(luck)])
 
-                    stats = (
-                        f"({att_space}{self.get_equipped_delta(current_equipped, item[1], 'att')} |"
-                        f"{cha_space}{self.get_equipped_delta(current_equipped, item[1], 'cha')} |"
-                        f"{int_space}{self.get_equipped_delta(current_equipped, item[1], 'int')} |"
-                        f"{dex_space}{self.get_equipped_delta(current_equipped, item[1], 'dex')} |"
-                        f"{luck_space}{self.get_equipped_delta(current_equipped, item[1], 'luck')} )"
-                    )
                 else:
-                    stats = (
-                        f"({att_space}{item[1].att if len(slot_name_org)  < 2 else item[1].att * 2} |"
-                        f"{cha_space}{item[1].cha if len(slot_name_org)  < 2 else item[1].cha * 2} |"
-                        f"{int_space}{item[1].int if len(slot_name_org) < 2 else item[1].int * 2} |"
-                        f"{dex_space}{item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2} |"
-                        f"{luck_space}{item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2} )"
+                    att = item[1].att if len(slot_name_org) < 2 else item[1].att * 2
+                    cha = item[1].cha if len(slot_name_org) < 2 else item[1].cha * 2
+                    int = item[1].int if len(slot_name_org) < 2 else item[1].int * 2
+                    dex = item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2
+                    luck = item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2
+                    rjuststat = max(
+                        [
+                            len(str(att)),
+                            len(str(cha)),
+                            len(str(int)),
+                            len(str(dex)),
+                            len(str(luck)),
+                        ]
                     )
+                rjuststat = 5
+                stats = (
+                    f"({att_space}{att:<{rjuststat}} |"
+                    f"{cha_space}{cha:<{rjuststat}} |"
+                    f"{int_space}{int:<{rjuststat}} |"
+                    f"{dex_space}{dex:<{rjuststat}} |"
+                    f"{luck_space}{luck:<{rjuststat}} )"
+                )
 
                 form_string += (
                     f"\n{str(item[1]):<{rjust}} - "
@@ -870,15 +883,15 @@ class Character(Item):
         return form_string + "\n"
 
     def get_equipped_delta(self, equiped: Item, to_compare: Item, stat_name: str) -> str:
-        if len(to_compare.slot) == 2 and len(equiped.slot) != 2:
+        if (equiped and len(equiped.slot) != 2) and (to_compare and len(to_compare.slot) == 2):
             equipped_left_stat = getattr(self.left, stat_name, 0)
             equipped_right_stat = getattr(self.right, stat_name, 0)
             equipped_stat = equipped_left_stat + equipped_right_stat
             comparing_to_stat = getattr(to_compare, stat_name, 0) * 2
-        elif len(equiped.slot) == 2 and len(to_compare.slot) != 2:
+        elif (equiped and len(equiped.slot) == 2) and (to_compare and len(to_compare.slot) != 2):
             equipped_stat = getattr(equiped, stat_name, 0) * 2
             comparing_to_stat = getattr(to_compare, stat_name, 0)
-        elif len(equiped.slot) == 2 and len(to_compare.slot) == 2:
+        elif (equiped and len(equiped.slot) == 2) and (to_compare and len(to_compare.slot) == 2):
             equipped_stat = getattr(equiped, stat_name, 0) * 2
             comparing_to_stat = getattr(to_compare, stat_name, 0) * 2
         else:
@@ -886,7 +899,7 @@ class Character(Item):
             comparing_to_stat = getattr(to_compare, stat_name, 0)
 
         diff = int(comparing_to_stat - equipped_stat)
-        return f"[{diff}]" if diff < 0 else f"{diff}"
+        return f"[{diff}]" if diff < 0 else f"+{diff}" if diff > 0 else "0"
 
     async def equip_item(self, item: Item, from_backpack: bool = True, dev=False):
         """This handles moving an item from backpack to equipment."""

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -487,14 +487,22 @@ class Character(Item):
 
     def remove_restrictions(self):
         if self.heroclass["name"] == "Ranger" and self.heroclass["pet"]:
+            requirements = (
+                PETS.get(self.heroclass["pet"]["name"], {}).get("bonuses", {}).get("req", {})
+            )
+            if "Ainz Ooal Gown" in self.sets and self.heroclass["pet"]["name"] in [
+                "Albedo",
+                "Rubedo",
+                "Guardians of Nazarick",
+            ]:
+                return
+
             if self.heroclass["pet"]["cha"] > (
                 self.total_cha + (self.total_int // 3) + (self.luck // 2)
             ):
                 self.heroclass["pet"] = {}
                 return
-            requirements = (
-                PETS.get(self.heroclass["pet"]["name"], {}).get("bonuses", {}).get("req", {})
-            )
+
             if requirements:
                 if requirements.get("set") and requirements.get("set") not in self.sets:
                     self.heroclass["pet"] = {}

--- a/adventure/charsheet.py
+++ b/adventure/charsheet.py
@@ -800,7 +800,9 @@ class Character(Item):
         )
         return final
 
-    async def get_backpack(self, forging: bool = False, consumed=None, rarity=None, slot=None, show_delta=False):
+    async def get_backpack(
+        self, forging: bool = False, consumed=None, rarity=None, slot=None, show_delta=False
+    ):
         if consumed is None:
             consumed = []
         bkpk = await self.get_sorted_backpack(self.backpack)
@@ -813,6 +815,7 @@ class Character(Item):
             slot_name_org = slot_group[0][1].slot
             slot_name = slot_name_org[0] if len(slot_name_org) < 2 else "two handed"
             form_string += f"\n\n {slot_name.title()} slot\n"
+            current_equipped = getattr(self, slot_name if slot != "two handed" else "left", None)
             async for item in AsyncIter(slot_group):
                 if forging and (item[1].rarity in ["forged", "set"] or item[1] in consumed_list):
                     continue
@@ -840,13 +843,22 @@ class Character(Item):
                     level = f"{e_level}"
 
                 if show_delta:
-                    pass
+
+                    stats = (
+                        f"({att_space}{self.get_equipped_delta(current_equipped, item[1], 'att')} |"
+                        f"{cha_space}{self.get_equipped_delta(current_equipped, item[1], 'cha')} |"
+                        f"{int_space}{self.get_equipped_delta(current_equipped, item[1], 'int')} |"
+                        f"{dex_space}{self.get_equipped_delta(current_equipped, item[1], 'dex')} |"
+                        f"{luck_space}{self.get_equipped_delta(current_equipped, item[1], 'luck')} )"
+                    )
                 else:
-                    stats = (f"({att_space}{item[1].att if len(slot_name_org)  < 2 else item[1].att * 2} |"
-                    f"{cha_space}{item[1].cha if len(slot_name_org)  < 2 else item[1].cha * 2} |"
-                    f"{int_space}{item[1].int if len(slot_name_org) < 2 else item[1].int * 2} |"
-                    f"{dex_space}{item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2} |"
-                    f"{luck_space}{item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2} )")
+                    stats = (
+                        f"({att_space}{item[1].att if len(slot_name_org)  < 2 else item[1].att * 2} |"
+                        f"{cha_space}{item[1].cha if len(slot_name_org)  < 2 else item[1].cha * 2} |"
+                        f"{int_space}{item[1].int if len(slot_name_org) < 2 else item[1].int * 2} |"
+                        f"{dex_space}{item[1].dex if len(slot_name_org) < 2 else item[1].dex * 2} |"
+                        f"{luck_space}{item[1].luck if len(slot_name_org) < 2 else item[1].luck * 2} )"
+                    )
 
                 form_string += (
                     f"\n{str(item[1]):<{rjust}} - "
@@ -856,6 +868,25 @@ class Character(Item):
                 )
 
         return form_string + "\n"
+
+    def get_equipped_delta(self, equiped: Item, to_compare: Item, stat_name: str) -> str:
+        if len(to_compare.slot) == 2 and len(equiped.slot) != 2:
+            equipped_left_stat = getattr(self.left, stat_name, 0)
+            equipped_right_stat = getattr(self.right, stat_name, 0)
+            equipped_stat = equipped_left_stat + equipped_right_stat
+            comparing_to_stat = getattr(to_compare, stat_name, 0) * 2
+        elif len(equiped.slot) == 2 and len(to_compare.slot) != 2:
+            equipped_stat = getattr(equiped, stat_name, 0) * 2
+            comparing_to_stat = getattr(to_compare, stat_name, 0)
+        elif len(equiped.slot) == 2 and len(to_compare.slot) == 2:
+            equipped_stat = getattr(equiped, stat_name, 0) * 2
+            comparing_to_stat = getattr(to_compare, stat_name, 0) * 2
+        else:
+            equipped_stat = getattr(equiped, stat_name, 0)
+            comparing_to_stat = getattr(to_compare, stat_name, 0)
+
+        diff = int(comparing_to_stat - equipped_stat)
+        return f"[{diff}]" if diff < 0 else f"{diff}"
 
     async def equip_item(self, item: Item, from_backpack: bool = True, dev=False):
         """This handles moving an item from backpack to equipment."""

